### PR TITLE
Snapshot tests now support a configuration file format to specify additional arguments.

### PR DESCRIPTION
**Pull request by Vighnesh-V**
This PR adds:
1) A configuration file format to snapshot tests - if you write a snap.config.luau file in the snapshot assets directory, we will look at that for additional args / per test overrides.
2) Ports the lint tests over and removes them from the corresponding directory
3) Runs these tests in CI so that we are dogfooding snapshot testing.

In the next series of PR's:
- style fixes - we ignored this chunk of the code base because this feature was way more experimental, but I'm going to do a separate pass that fixes the casing here.
- documentation about snapshot testing
- Separate out stderr and stdout as separate assets.
- porting more tests

### CODE REVIEW

File: lute/cli/commands/test/snap/runner.luau [hunk 1]
```diff
@@ -47,10 +97,33 @@ function runner.runsnapshots(files: { path.Path }): snapty.snapresult
...
-		local artifactpath = artifactpath(filepath)
+		local configuration: snapty.SnapshotConfiguration? = nil
+		local directory = path.dirname(filepath)
+		local maybeConfig = configCache[directory]
+		if maybeConfig == nil then
+			local ok, conf = pcall(retrieveConfiguration, directory)
+			if not ok then
+				-- Something went wrong loading this module
+				break
+			end
```
Comment (annieetang) [inline lute/cli/commands/test/snap/runner.luau:133]: do we want to print/warn here?

**Review by annieetang — APPROVED** [review]
looks good to me thanks for adding additional support! just one edge case q
will also defer to @skanosue for linting stuff

File: lute/cli/commands/test/snap/runner.luau (hunk 1 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/runner.luau:133]: I could add a print statement if we're not able to retrieve configuration here.

File: lute/cli/commands/test/snap/types.luau [hunk 2]
```diff
@@ -26,4 +26,18 @@ export type snapresult = {
 	new: { snapnew },
 }
 
+--- Configuration for an individual test
+export type TestConfiguration = {
+	--- Arguments to pass `lute` before the snapshot file
+	args: { string },
```
Comment (skberkeley) [inline lute/cli/commands/test/snap/types.luau]: if we want external users to adopt snapshot testing maybe we don't prefix `lute` every time?

File: lute/cli/commands/test/snap/runner.luau [hunk 3]
```diff
@@ -1,25 +1,75 @@
+local fs = require("@std/fs")
 local process = require("@std/process")
 local path = require("@std/path")
-local fs = require("@std/fs")
-local difftext = require("@batteries/difftext")
+local luau = require("@std/luau")
```
Comment (skberkeley) [inline lute/cli/commands/test/snap/runner.luau]: l < p

File: lute/cli/commands/test/snap/runner.luau (hunk 1 above)
Comment (annieetang) [reply lute/cli/commands/test/snap/runner.luau:133]: sgtm! would be nice for logging

File: lute/cli/commands/test/snap/runner.luau [hunk 4]
```diff
@@ -63,7 +136,8 @@ function runner.runsnapshots(files: { path.Path }): snapty.snapresult
 			continue
 		end
 
-		local expected = fs.readFileToString(artifactpath)
+		local rawExpected = fs.readFileToString(artifactpath)
+		local expected, _ = string.gsub(rawExpected, "\r\n", "\n")
```
Comment (skberkeley) [inline lute/cli/commands/test/snap/runner.luau]: is this bc we run ci on windows?

File: lute/cli/commands/test/snap/runner.luau [hunk 5]
```diff
@@ -1,25 +1,75 @@
...
+
+local function getCommandArgs(
+	execpath: string,
+	filepath: string,
+	configuration: snapty.SnapshotConfiguration?,
+	artifactPath: string
+)
+	local args = { execpath }
+	local testname = strext.removeSuffix(artifactPath, ".snap")
+	if not configuration then
```
Comment (skberkeley) [inline lute/cli/commands/test/snap/runner.luau:50]: can you add a comment here or something? it's just a little confusing having two variables names `config` and `configuration`

**Review by skberkeley — CHANGES_REQUESTED** [review]
talked about this offline, but dym laying what the concrete plan is for checking exit codes? otherwise we're loosening the lint tests you ported

File: lute/cli/commands/test/snap/runner.luau (hunk 4 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/runner.luau]: yup

File: lute/cli/commands/test/snap/runner.luau (hunk 5 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/runner.luau:50]: I'll annotate this a bit!

File: lute/cli/commands/test/snap/runner.luau (hunk 3 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/runner.luau]: will fix!

File: lute/cli/commands/test/snap/runner.luau (hunk 4 above)
Comment (annieetang) [reply lute/cli/commands/test/snap/runner.luau]: is having a helper function for this useful? no strong opinions but i do see you use `string.gsub(..., "\r\n", "\n")` twice

File: lute/cli/commands/test/snap/types.luau (hunk 2 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/types.luau]: The intention here is explicitly that users use `luau` to write scripts that produce output that they could snapshot against. The prior art here is expect tests in dune - you can write ocaml code that writes to stdout, and that gets captured inline.

**Comment by Vighnesh-V** [comment]
@skberkeley Discussed offline, but the plan here is to make the expected exit code configurable. I've added it as a configuration option.

File: lute/cli/commands/test/snap/runner.luau (hunk 4 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/runner.luau]: I think I can just fold this into normalize snapshot output and omit this use.

File: lute/cli/commands/test/snap/runner.luau (hunk 5 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/runner.luau:50]: instead of this, I've tweaked this code so it's less of an issue. I think it makes sense to rename the `config` property to `default` for readability.

File: lute/cli/commands/test/snap/runner.luau [hunk 6]
```diff
@@ -1,7 +1,7 @@
 local fs = require("@std/fs")
+local luau = require("@std/luau")
 local process = require("@std/process")
 local path = require("@std/path")
```
Comment (skberkeley) [inline lute/cli/commands/test/snap/runner.luau]: oh wait i missed `path < process` sorry

File: lute/cli/commands/test/snap/runner.luau [hunk 7]
```diff
@@ -63,9 +165,13 @@ function runner.runsnapshots(files: { path.Path }): snapty.snapresult
 			continue
 		end
 
-		local expected = fs.readFileToString(artifactpath)
+		local rawExpected = fs.readFileToString(artifactpath)
+		local expected = normalizeSnapshotOutput(rawExpected, cwd)
+		-- A test passes when a) no exit code checking is requested and b) actual == expected
+		-- It also passes when a) exit code checking is request and b) actual == expected
```
Comment (skberkeley) [inline lute/cli/commands/test/snap/runner.luau]: this comment seems wrong?
shouldn't it just be 
a) no exit code checking is requested or b) actual == expected

**Review by skberkeley — APPROVED** [review]

File: lute/cli/commands/test/snap/runner.luau (hunk 7 above)
Comment (Vighnesh-V) [reply lute/cli/commands/test/snap/runner.luau]: Comment is wrong on the second line. Should be:
1. no exit code checking is requested -> check just the snapshot contents
2. exit code checking is requested -> check snapshot contents AND exit code.
